cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/MacroParser.java

```diff
@@ -0,0 +1,267 @@
+   * @throws InvalidMacroException on invalid macro syntax.
+   * @return if str contains a macro.
+   */
+  public boolean containsMacro(String str) throws InvalidMacroException {
+    // used to check macro syntax
+    substituteMacro(str, 0, new PropertyMacros(), false, false);
+    return (findRightmostMacro(str, null, false, false) != null);
+  }
+
+  /**
+   * Creates a set of all macro arguments present in a string that a user must specify to expand the string.
+   * TODO: Get this working for REST API
+   * @param str the raw string containing macro syntax.
+   * @throws InvalidMacroException on invalid macro syntax
+   * @return a set of all macros present in str.
+   */
+  public PropertyMacros getMacros(String str) throws InvalidMacroException {
+    PropertyMacros propertyMacros = new PropertyMacros();
+    substituteMacro(str, 0, propertyMacros, false, true);
+    return propertyMacros;
+  }
+
+  /**
+   * Substitutes the provided string with a given macro evaluator. Expands macros from right-to-left recursively.
+   * @param str the raw string containing macro syntax
+   * @return the original string with all macros expanded and substituted
+   */
+  public String substituteMacro(String str) {
+    // final string has escapes not directly embedded in macro syntax replaced
+    return replaceEscapedSyntax(substituteMacro(str, 0, new PropertyMacros(), true, false));
+  }
+
+  private String substituteMacro(String str, int depth, PropertyMacros propertyMacros, boolean substitute,
+                                 boolean surfaceMacros)
+    throws InvalidMacroException {
+    if (depth > MAX_SUBSTITUTION_DEPTH) {
+      throw new InvalidMacroException(String.format("Failed substituting maco '%s', expansion exceeded %d levels.",
+                                                    str, MAX_SUBSTITUTION_DEPTH));
+    }
+
+    MacroMetadata macroPosition = findRightmostMacro(str, propertyMacros, substitute, surfaceMacros);
+    while (macroPosition != null) {
+      str = str.substring(0, macroPosition.startIndex) +
+        substituteMacro(macroPosition.substitution, depth + 1, propertyMacros, substitute, surfaceMacros) +
+        str.substring(macroPosition.endIndex + 1);
+      macroPosition = findRightmostMacro(str, propertyMacros, substitute, surfaceMacros);
+    }
+    return str;
+  }
+
+  /**
+   * Find the rightmost macro in the specified string. If no macro is found, returns null.
+   *
+   * @param str the string to find a macro in
+   * @param propertyMacros a set of arguments to surface if called from getMacros
+   * @param substitute whether to perform macro substitution or not
+   * @return the rightmost macro and its position in the string
+   * @throws InvalidMacroException if invalid macro syntax was found.
+   */
+  @Nullable
+  private MacroMetadata findRightmostMacro(String str, PropertyMacros propertyMacros, boolean substitute,
+                                           boolean surfaceMacros)
```

Comment:
Yeah, this was a temporary implementation while I was moving stuff around in MacroParser. This will all go with the DummyEvaluator we talked about today.